File: src/main/java/monnef/core/utils/RegistryUtils.java
/*
 * Jaffas and more!
 * author: monnef
 */

package monnef.core.utils;

import com.google.common.collect.Multimap;
import cpw.mods.fml.common.Loader;
import cpw.mods.fml.common.LoaderState;
import cpw.mods.fml.common.ModContainer;
import cpw.mods.fml.common.registry.BlockProxy;
import cpw.mods.fml.common.registry.GameRegistry;
import cpw.mods.fml.common.registry.LanguageRegistry;
import cpw.mods.fml.relauncher.ReflectionHelper;
import monnef.core.MonnefCorePlugin;
import monnef.core.api.IItemBlock;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

import static monnef.core.MonnefCorePlugin.Log;


public class RegistryUtils {
    public static void registerBlock(Block block) {
        GameRegistry.registerBlock(block, block.getUnlocalizedName());
    }

    public static void registerBlock(Block block, String title) {
        registerBlock(block);
        LanguageRegistry.addName(block, title);
    }

    public static void registerBlock(Block block, String name, String title) {
        block.setBlockName(name);
        registerBlock(block, title);
    }

    public static void registerMultiBlock(Block block, Class<?> itemBlock, String[] titles) {
        registerMultiBlock(block, itemBlock, titles, null);
    }

    public static void registerMultiBlock(Block block, Class<?> itemBlock, String name, String title) {
        registerMultiBlock(block, itemBlock, new String[]{title}, new String[]{name});
    }

    public static void registerMultiBlock(Block block, Class<?> itemBlock, String[] titles, String[] names) {
        if (block.getUnlocalizedName() == null && (names == null || names.length <= 0)) {
            throw new RuntimeException("Cannot find a block name - " + itemBlock.getSimpleName() + ".");
        }

        Class<?> cls = itemBlock;
        if (ItemBlock.class.isAssignableFrom(cls)) {
            if (names != null && names.length != 1) {
                throw new RuntimeException("cannot have multiple custom names for ItemBlock");
            }
            block.setBlockName(names[0]);
            GameRegistry.registerBlock(block, (Class<? extends ItemBlock>) itemBlock, block.getUnlocalizedName());
        } else if (IItemBlock.class.isAssignableFrom(cls)) {
            registerMyBlock(block, (Class<? extends IItemBlock>) itemBlock, block.getUnlocalizedName(), names);
        } else {
            throw new RuntimeException("Unknown class in block registration.");
        }

        registerSubBlocks(block, titles);
    }

    private static void registerMyBlock(Block block, Class<? extends IItemBlock> itemclass, String blockName, String[] names) {
        // heavily based on GameRegistry.registerBlock of Forge
        if (Loader.instance().isInState(LoaderState.CONSTRUCTING)) {
            Log.printWarning("Registering block in non-constructing state!");
        }
        try {
            assert block != null : "registerBlock: block cannot be null";
            assert itemclass != null : "registerBlock: itemclass cannot be null";
            int blockItemId = block.blockID - 256;
            Constructor<? extends IItemBlock> itemCtor;
            Item i;
            try {
                itemCtor = itemclass.getConstructor(int.class);
                i = (Item) itemCtor.newInstance(blockItemId);
            } catch (NoSuchMethodException e) {
                itemCtor = itemclass.getConstructor(int.class, Block.class);
                i = (Item) itemCtor.newInstance(blockItemId, block);
            }
            if (names != null) {
                ((IItemBlock) i).setSubNames(names);
            }
            GameRegistry.registerItem(i, blockName, null);
        } catch (Exception e) {
            Log.printSevere("Problem in registerMyBlock.");
            throw new RuntimeException(e);
        }

        Field br = ReflectionHelper.findField(GameRegistry.class, "blockRegistry");
        Multimap<ModContainer, BlockProxy> blockRegistry;
        try {
            blockRegistry = (Multimap<ModContainer, BlockProxy>) br.get(null);
        } catch (IllegalAccessException e) {
            Log.printSevere("Problem in registerMyBlock.");
            throw new RuntimeException(e);
        }
        blockRegistry.put(Loader.instance().activeModContainer(), (BlockProxy) block);
    }

    private static void registerSubBlocks(Block block, String[] names) {
        for (int ix = 0; ix < names.length; ix++) {
            ItemStack multiBlockStack = new ItemStack(block, 1, ix);
            LanguageRegistry.addName(multiBlockStack, names[multiBlockStack.getItemDamage()]);
        }
    }

    public static void registerSubItems(Item item, String[] names) {
        for (int ix = 0; ix < names.length; ix++) {
            ItemStack multiBlockStack = new ItemStack(item, 1, ix);
            LanguageRegistry.addName(multiBlockStack, names[multiBlockStack.getItemDamage()]);
        }
    }

    public static <T extends Item> T registerItem(T item, String name, String title) {
        item.setUnlocalizedName(name);
        GameRegistry.registerItem(item, item.getUnlocalizedName());
        LanguageRegistry.addName(item, title);
        return item;
    }

    public static String getTitle(ItemStack stack) {
        String title = stack.getDisplayName();
        if (title == null || title.isEmpty() || title.contains("item.") || title.contains("tile.") || title.contains(".name")) {
            title = LanguageRegistry.instance().getStringLocalization(stack.getItem().getUnlocalizedName() + ".name");
        }
        if (title == null || title.isEmpty()) {
            title = stack.getUnlocalizedName();
            MonnefCorePlugin.Log.printWarning(String.format("Unable to find translation of %s.", title));
        }
        return title;
    }

    public static void registerBlockPackingRecipe(ItemStack input, ItemStack outputBlock) {
        ItemStack[] inputItems = new ItemStack[9];
        for (int i = 0; i < 9; i++) {
            inputItems[i] = input.copy();
        }

        GameRegistry.addShapelessRecipe(outputBlock.copy(), inputItems);
    }
}

File: src/main/scala/monnef/core/utils/RegistryUtils.scala
package monnef.core.utils

import net.minecraft.block.Block
import cpw.mods.fml.common.registry.{FMLControlledNamespacedRegistry, GameData, LanguageRegistry, GameRegistry}
import net.minecraft.item.{ItemStack, Item, ItemBlock}
import monnef.core.api.IItemBlock
import cpw.mods.fml.common.{FMLCommonHandler, FMLLog, LoaderState, Loader}
import monnef.core.MonnefCorePlugin._
import java.lang.reflect.Constructor
import cpw.mods.fml.relauncher.ReflectionHelper
import monnef.core.MonnefCorePlugin
import net.minecraft.client.resources.I18n
import scalautils._
import net.minecraft.util.StatCollector

object RegistryUtils {
  def registerBlock(block: Block) {
    GameRegistry.registerBlock(block, block.getUnlocalizedName)
  }

  def registerBlockWithName(block: Block, name: String) {
    block.setBlockName(name)
    registerBlock(block)
  }

  def registerMultiBlock(block: Block, itemBlock: Class[_], name: String) {
    registerMultiBlock(block, itemBlock, Array[String](name))
  }

  def registerMultiBlock(block: Block, itemBlock: Class[_], names: Array[String]) {
    if (block.getUnlocalizedName == null && (names == null || names.length <= 0)) {
      throw new RuntimeException("Cannot find a block name - " + itemBlock.getSimpleName + ".")
    }
    names.foreach { case name => if (name.contains(" ") || name(0).toLower != name(0)) Log.printWarning(s"Invalid name '$name' used for registration of multi block ${block.getUnlocalizedName} (${block.getClass}})")}
    val cls: Class[_] = itemBlock
    if (classOf[ItemBlock].isAssignableFrom(cls)) {
      if (names != null && names.length != 1) {
        throw new RuntimeException("cannot have multiple custom names for ItemBlock")
      }
      block.setBlockName(names(0))
      GameRegistry.registerBlock(block, itemBlock.asInstanceOf[Class[_ <: ItemBlock]], block.getUnlocalizedName)
    }
    else if (classOf[IItemBlock].isAssignableFrom(cls)) {
      registerMyBlock(block, itemBlock.asInstanceOf[Class[_ <: IItemBlock]], block.getUnlocalizedName, names)
    }
    else {
      throw new RuntimeException("Unknown class in block registration.")
    }
  }

  private def registerMyBlock(block: Block, itemclass: Class[_ <: IItemBlock], blockName: String, names: Array[String]) {
    // heavily based on GameRegistry.registerBlock of Forge
    if (Loader.instance.isInState(LoaderState.CONSTRUCTING)) {
      Log.printWarning("Registering block in non-constructing state!")
    }
    var i: Item = null
    try {
      assert(block != null, "registerBlock: block cannot be null")
      assert(itemclass != null, "registerBlock: itemclass cannot be null")

      try {
        var itemCtor: Constructor[_ <: IItemBlock] = null
        itemCtor = itemclass.getConstructor(classOf[Block])
        i = itemCtor.newInstance(block).asInstanceOf[Item]
      } catch {
        case _: NoSuchMethodException =>
          try {
            i = itemclass.newInstance().asInstanceOf[Item]
          } catch {
            case e: NoSuchMethodException =>
              throw new RuntimeException("Block doesn't have suitable constructor (either parameter-less or accepting a block).", e)
          }
      }

      if (names != null) {
        i.asInstanceOf[IItemBlock].setSubNames(names)
      }

      try {
        //GameData.getMain.registerBlock(block, name,)
        GameDataAccessor.registerBlock(block, blockName)
      } catch {
        case e: Exception =>
          Log.printSevere("Problem in registerMyBlock - Forge block registration")
          throw new RuntimeException(e)
      }

      //GameRegistry.registerItem(i, blockName, null)
      registerCustomItemBlock(i, blockName)
    }
    catch {
      case e: Exception =>
        Log.printSevere("Problem in registerMyBlock.")
        throw new RuntimeException(e)
    }
  }

  private def registerCustomItemBlock(item: Item, name: String): Int = {
    val block = item.asInstanceOf[IItemBlock].getBlock
    val idHint: Integer = GameData.getBlockRegistry.getId(block)
    if (idHint == -1) {
      throw new RuntimeException("Block must be registered first.")
    }
    else {
      FMLLog.fine("[registerCustomItemBlock] Found matching Block %s for ItemBlock %s at id %d", block, item, idHint)
      GameDataAccessor.freeSlot(idHint, item)
      if (GameData.getItemRegistry.getObjectById(idHint) != null) throw new RuntimeException("freeSlot was unsuccessful")
    }

    val itemId: Int = GameDataAccessor.iItemRegistryAdd(idHint, name, item, GameDataAccessor.availabilityMap)
    if (itemId != idHint) throw new IllegalStateException("Block -> ItemBlock insertion failed.")
    GameDataAccessor.verifyCustomItemBlockName(item.asInstanceOf[IItemBlock])
    GameDataAccessor.useSlot(itemId)
    itemId
  }

  def registerItem[T <: Item](item: T): T = {
    val name = item.getUnlocalizedName
    if ("item.null".equals(name)) throw new RuntimeException("Trying to register not named item, this will cause unpredictable errors if not fixed.")
    GameRegistry.registerItem(item, name)
    item
  }

  def registerItem[T <: Item](item: T, name: String): T = {
    item.setUnlocalizedName(name)
    registerItem(item)
  }

  private val titleGetters: Seq[(ItemStack) => (String, String)] = Seq(
    s => "D" -> s.getDisplayName,
    s => "s" -> StatCollector.translateToLocal(s.getUnlocalizedName),
    s => "S" -> StatCollector.translateToLocal(s.getUnlocalizedName + ".name"),
    s => "i" -> I18n.format(s.getUnlocalizedName),
    s => "I" -> I18n.format(s.getUnlocalizedName + ".name"),
    s => "l" -> (try {LanguageRegistry.instance().getStringLocalization(s.getUnlocalizedName)} catch {case _: Throwable => null}),
    s => "L" -> (try {LanguageRegistry.instance().getStringLocalization(s.getUnlocalizedName + ".name")} catch {case _: Throwable => null}),
    s => "u" -> LanguageRegistry.instance().getStringLocalization(s.getUnlocalizedName, "en_US"),
    s => "U" -> LanguageRegistry.instance().getStringLocalization(s.getUnlocalizedName + ".name", "en_US"),

    s => "?" -> ("[???]" + s.getUnlocalizedName)
  )

  def getTitle(stack: ItemStack): String = {
    if (stack == null) throw new RuntimeException("Null stack.")
    if (stack.getItem == null) throw new RuntimeException("Null item in stack.")
    def seemsInvalid(name: String): Boolean = name == null || "".equals(name) || name.startsWith("item.") || name.startsWith("tile.") || name.endsWith(".name")

    titleGetters.foldLeft(("", "")) { case (acc, f) => if (seemsInvalid(acc._2)) f(stack) else acc} |> { case (id, t) => (if (MonnefCorePlugin.debugEnv) id else "") + t}
  }

  def registerBlockPackingRecipe(input: ItemStack, outputBlock: ItemStack) {
    if (input == null || input.getItem == null || outputBlock == null || outputBlock.getItem == null) {
      throw new NullPointerException
    }
    if (Item.getIdFromItem(input.getItem) == 0) {
      throw new RuntimeException(s"Input ${input.getItem.getUnlocalizedName} has invalid ID.")
    }
    if (Item.getIdFromItem(outputBlock.getItem) == 0) {
      throw new RuntimeException(s"Output ${outputBlock.getItem.getUnlocalizedName} has invalid ID.")
    }
    val inputItems: Array[ItemStack] = Array.fill(9) {input.copy}
    GameRegistry.addShapelessRecipe(outputBlock.copy, inputItems: _*)
  }
}
